test/k8s/res_limits_test.go

```diff
@@ -0,0 +1,512 @@
+
+	BeforeAll(func() {
+		opts := []KumaDeploymentOption{
+			WithCtlOpts(map[string]string{
+				"--set": "" +
+					"kuma.controlPlane.resources.requests.cpu=500m," +
+					"kuma.controlPlane.resources.requests.memory=512Mi," +
+					"kuma.controlPlane.resources.limits.cpu=2000m," +
+					"kuma.controlPlane.resources.limits.memory=2048Mi",
+				"--env-var": "" +
+					"KUMA_RUNTIME_KUBERNETES_LEADER_ELECTION_LEASE_DURATION=100s," +
+					"KUMA_RUNTIME_KUBERNETES_LEADER_ELECTION_RENEW_DEADLINE=80s",
+			}),
+		}
+
+		alternativeContainerRegistry, _ = os.LookupEnv("ALTERNATIVE_CONTAINER_REGISTRY")
+
+		if alternativeContainerRegistry != "" {
+			opts = append(opts,
+				WithCtlOpts(map[string]string{
+					"--dataplane-registry": alternativeContainerRegistry,
+				}))
+		}
+
+		opts = append(opts,
+			WithCtlOpts(map[string]string{
+				"--license-path": kmeshLicense,
+			}))
+
+		err := NewClusterSetup().
+			Install(Kuma(core.Zone, opts...)).
+			Setup(cluster)
+		Expect(err).ToNot(HaveOccurred())
+	})
+
+	BeforeEach(func() {
+		Expect(framework.ReportSpecStart(cluster)).To(Succeed())
+	})
+
+	E2EAfterAll(func() {
+		Expect(cluster.TriggerDeleteNamespace(TestNamespace)).To(Succeed())
+		Expect(cluster.DeleteKuma()).To(Succeed())
+	})
+
+	It("should deploy the mesh", func() {
+		// just to see stabilized stats before we go further
+		Expect(true).To(BeTrue())
+	})
+
+	It("should deploy mesh wide policy", func() {
+		policy := `
+apiVersion: kuma.io/v1alpha1
+kind: MeshRateLimit
+metadata:
+  name: mesh-rate-limit
+  namespace: kong-mesh-system
+spec:
+  targetRef:
+    kind: Mesh
+    proxyTypes:
+      - Sidecar
+  from:
+    - targetRef:
+        kind: Mesh
+      default:
+        local:
+          http:
+            requestRate:
+              num: 10000
+              interval: 1s
+            onRateLimit:
+              status: 429
+`
+		Expect(cluster.Install(YamlK8s(policy))).To(Succeed())
+	})
+
+	Context("load", func() {
+		var cpu int
+		var memory int
+		var numServices = 5
+		var instancesPerService = 1
+		var goDebugAdded = false
+		var goMemLimitAdded = false
+
+		adjustResource := func(miliCPU, memMega int, addGoMemLimitEnv bool) {
+			printLog("adjusting control plane resource limits to cpu %dm, memory %dMi\n", miliCPU, memMega)
+
+			patchJson := `[
+{"op": "replace", "path": "/spec/template/spec/containers/0/resources/requests/cpu", "value": "%dm"},
+{"op": "replace", "path": "/spec/template/spec/containers/0/resources/requests/memory", "value": "%dMi"},
+{"op": "replace", "path": "/spec/template/spec/containers/0/resources/limits/cpu", "value": "%dm"},
+{"op": "replace", "path": "/spec/template/spec/containers/0/resources/limits/memory", "value": "%dMi"}`
+
+			if !goDebugAdded {
+				goDebugAdded = true
+				patchJson += `,{"op": "add", "path": "/spec/template/spec/containers/0/env/-", "value": {"name": "GODEBUG", "value":"gctrace=1"}}`
+			}
+
+			if addGoMemLimitEnv {
+				memLimit := memMega - 50
+				if memLimit < 10 {
```

Comment:
Can it be lower than 10? Why 10 is minimum?

Reply:
It's a basic protection to the value making sure it's not too low. I'll use 90% instead.